### PR TITLE
Support model type in model spec

### PR DESCRIPTION
**Pull request by eddyxu**
This PR sets a rule to find a "ModelType" class via `MODEL_TYPE` attribute.
It points to a class "ModelSpec" which wraps the actual model (pytorch), and acts as a proxy. It offers the flexibility to adapt non-standard pytorch models.

### CODE REVIEW

File: python/rikai/contrib/torch/transforms/ssd.py [hunk 1]
```diff
@@ -24,12 +24,7 @@
 
 
 def pre_processing(options: Dict[str, Any]) -> Callable:
-    return T.Compose(
-        [
-            T.Resize([300, 300]),
-            T.ToTensor(),
-        ]
-    )
+    return T.ToTensor()
```
Comment (eddyxu) [inline python/rikai/contrib/torch/transforms/ssd.py:27]: SSD model will do resize in the model

**Comment by eddyxu** [comment]
Related to https://github.com/eto-ai/rikai/issues/520

File: python/rikai/contrib/torch/detections.py [hunk 2]
```diff
@@ -14,4 +14,4 @@
 
 
 # Object detection output schema
-OUTPUT_SCHEMA = "array<struct<box:box2d, score:float, label_id:int>>"
+OUTPUT_SCHEMA = "array<struct<box:box2d, score:float, label:int>>"
```
Comment (darcy-shen) [inline python/rikai/contrib/torch/detections.py:17]: label_id is the id of label, label should be the text of the label

I think we should use label_id here.

File: python/rikai/contrib/torch/transforms/ssd.py (hunk 1 above)
Comment (darcy-shen) [reply python/rikai/contrib/torch/transforms/ssd.py:27]: if we do resizing in pre_processing, the DataLoader will load it with resizing in parallel.

Is the target width or height fixed?

Comment (darcy-shen) [reply python/rikai/contrib/torch/transforms/ssd.py:27]: If it is fixed for SSD, and it is (300, 300)，why not do benchmark and find the best performant approach?

File: src/main/scala/ai/eto/rikai/sql/spark/parser/RikaiExtAstBuilder.scala [hunk 3]
```diff
@@ -54,6 +54,10 @@ private[parser] class RikaiExtAstBuilder
           case _         => None
         }
     }
+    val modelType: Option[String] = ctx.modeltype match {
+      case null => None
+      case _    => Option(ctx.modeltype.getText)
```
[the remark this answers is not in the document]
Comment (darcy-shen) [reply src/main/scala/ai/eto/rikai/sql/spark/parser/RikaiExtAstBuilder.scala:59]: Option(ctx.modelType).map(_.getText)

File: python/rikai/spark/sql/model.py [hunk 4]
```diff
@@ -0,0 +1,260 @@
...
+
+    @abstractmethod
+    def schema(self) -> str:
+        pass
+
+    def dataType(self) -> "pyspark.sql.types.DataType":
+        return parse_schema(self.schema())
+
+    @abstractmethod
+    def transform(self) -> Callable:
```
Comment (darcy-shen) [inline python/rikai/spark/sql/model.py:215]: it seems transform is `pre_processing` and predict is `post_processing`, could you add some comments here?

File: python/rikai/pytorch/models/ssd.py [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+
+from .torchvision import ObjectDetectionModelType
+
+__all__ = ["MODEL_TYPE"]
+
+MODEL_TYPE = ObjectDetectionModelType("ssd")
```
[the remark this answers is not in the document]
Comment (darcy-shen) [reply python/rikai/pytorch/models/ssd.py:19]: Where is the name actually used?

File: python/rikai/spark/sql/model.py [hunk 6]
```diff
@@ -0,0 +1,261 @@
...
+        except ModuleNotFoundError:
+            pass
+    else:
+        raise ModuleNotFoundError(
+            f"Model spec not found for model: {model_type}/{flavor}"
+        )
+
+
+class SpecPayload(ABC):
+    """Model Spec Payload
```
[the remark this answers is not in the document]
Comment (eddyxu) [reply python/rikai/spark/sql/model.py:97]: This is used to pass ModelSpec between jvm and python

File: python/rikai/contrib/torch/transforms/ssd.py (hunk 1 above)
Comment (eddyxu) [reply python/rikai/contrib/torch/transforms/ssd.py:27]: In ssd model, the resize can be done in GPU. 

But the purpose of removing it was not due to a concern of performance.  Letting SSD to handle it (better) can allow us to go through Coco dataset with different size and different number of channels of images.  The current `main` branch (before #529) can not do batch_size > 1 on a full coco dataset scan.

File: python/rikai/spark/sql/model.py [hunk 7]
```diff
@@ -0,0 +1,260 @@
...
+        pass
+
+    @abstractmethod
+    def predict(self, *args, **kwargs) -> Any:
+        pass
+
+    def __call__(self, *args, **kwargs) -> Any:
+        return self.predict(*args, **kwargs)
+
+    def release(self):
```
[the remark this answers is not in the document]
Comment (eddyxu) [reply python/rikai/spark/sql/model.py:225]: Does not need to be, it can be just a noop by default?
I will just add a comment here

File: python/rikai/pytorch/models/ssd.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (eddyxu) [reply python/rikai/pytorch/models/ssd.py:19]: ok i will change this to 

```class SSDModelType(ObjectDetectionModelType)
          def __init__():

MODEL_TYPE = SSDModelType()
```

[the remark this answers is not in the document]
Comment (eddyxu) [reply python/rikai/pytorch/models/ssd.py:19]: Added the name to `__repr__` for logging and debugging purpose.

File: src/main/scala/ai/eto/rikai/sql/spark/parser/RikaiExtAstBuilder.scala (hunk 3 above)
[the remark this answers is not in the document]
Comment (eddyxu) [reply src/main/scala/ai/eto/rikai/sql/spark/parser/RikaiExtAstBuilder.scala:59]: done

File: python/rikai/contrib/torch/detections.py (hunk 2 above)
Comment (eddyxu) [reply python/rikai/contrib/torch/detections.py:17]: ok,

File: python/rikai/spark/sql/model.py (hunk 4 above)
Comment (eddyxu) [reply python/rikai/spark/sql/model.py:215]: done

File: python/rikai/pytorch/models/torchvision.py [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+DEFAULT_MIN_SCORE = 0.5
+
+
+class ObjectDetectionModelType(ModelType):
+    """Shared ModelSpec for object detections in Torchvision
+
+    https://pytorch.org/vision/stable/models.html
+    """
+
+    def __init__(self, name: str):
```
[the remark this answers is not in the document]
Comment (eddyxu) [reply python/rikai/pytorch/models/torchvision.py:39]: added to the `__repr__`

**Review by changhiskhan — APPROVED** [review]
i think this is a step-forward in that it provides a more consistent interface for adding new models. However, it seems to make the process of adding a new model more complicated. At least worth documenting what a rikai dev have to do to add a new model. Possible consider building something on top of this extraction to simplify things

**Comment by eddyxu** [comment]
Ok, filed #533 and #532 as follow ups
